hkmc2/shared/src/test/mlscript-compile/apps/parsing-web-demo/main.mls

```diff
@@ -0,0 +1,223 @@
+import "../parsing/Parser.mls"
+import "../parsing/Lexer.mls"
+import "../../Iter.mls"
+import "../../XML.mls"
+import "../../Option.mls"
+import "../../Runtime.mls"
+import "../../Predef.mls"
+import "../parsing/TreeHelpers.mls"
+import "../parsing/Extension.mls"
+import "../parsing/ParseRuleVisualizer.mls"
+import "../parsing/Rules.mls"
+import "../parsing/vendors/railroad/railroad.mjs"
+
+open XML { elem }
+open Predef
+open Option { Some, None }
+
+declare val document
+declare val customElements
+declare class HTMLElement
+
+module Main with ...
+
+let query = document.querySelector.bind(document)
+let editor = query("#editor")
+let selector = query("select#example")
+let parseButton = query("button#parse")
+let outputPanel = query("#output")
+
+fun showExample1() =
+  set editor.value = """let spaces n = make_string n " ";;
+let disk size =
+    let right_half = make_string size ">"
+    and left_half = make_string size "<"
+    in left_half ^ "|" ^ right_half;;
+let disk_number n largest_disk_size =
+    let white_part = spaces (largest_disk_size + 1 - n) in
+    white_part ^ (disk n) ^ white_part;;
+let peg_base largest_disk_size =
+    let half = make_string largest_disk_size "_" in
+    " " ^ half ^ "|" ^ half ^ " ";;
+let rec peg largest_disk_size = function
+  | (0, []) -> []
+  | (0, head::rest) ->
+      disk_number head largest_disk_size ::
+      peg largest_disk_size (0, rest)
+  | (offset, lst) ->
+      disk_number 0 largest_disk_size ::
+      peg largest_disk_size (offset-1, lst);;
+let rec join_lines l1 l2 l3 =
+  match (l1, l2, l3) with
+  | ([], [], []) -> []
+  | (t1::r1, t2::r2, t3::r3) -> (t1 ^ t2 ^ t3) :: join_lines r1 r2 r3
+  | _ -> failwith "join_lines";;
+"""
+
+fun showExample2() =
+  set editor.value = """#newKeyword ("hello", Some 3, Some 3)
+#newKeyword ("goodbye", None, None)
+  
+#newCategory("greeting")
+  
+#extendCategory("greeting", [ keyword("hello"), "term", "greeting" ], foo)
+#extendCategory("greeting", [ keyword("goodbye") ], bar)
+  
+#extendCategory("decl", [ "greeting" ], baz)
+  
+  
+hello "Rob" hello "Bob" goodbye
+  
+#diagram ""
+"""
+
+showExample1()
+
+editor.addEventListener of "keydown", event =>
+  if event.key is "Tab" do
+    event.preventDefault()
+    let start = editor.selectionStart
+    let end = editor.selectionEnd
+    set editor.value = editor.value.substring(0, start) + "  " + editor.value.substring(end)
+    set editor.selectionEnd = start + 2
+    set editor.selectionStart = editor.selectionEnd
+
+selector.addEventListener of "change", event => if selector.value is
+  "example-1" do showExample1()
+  "example-2" do showExample2()
```

Comment:
This inexhaustive match looks a bit uncomfortable. Couldn't we set all examples, along with their names and titles, in one place, and then use this info in the relevant places accordingly? The HTML `select` list could start with no `option` and would be updated by the JS based on this single source of truth.

Reply:
Done in 828589c.